### ISSUE TITLE
The put command sometimes result in broken templates on the server

### ISSUE THREAD

**Issue by wna-se** [ds-wizard/dsw-tdk#13]
We use the docker compose setup of the wizard to work on our templates locally as follows:

1. download the template from a shared server
2. edit the template files locally 
3. upload to the local server running in Docker
4. repeat from 2 until we're happy
5. upload to the shared server

So far we have noticed that `dsw-tdk put` command sometimes result in broken templates on the server (templates that consistently fail to render any documents) without any changes made to the template files. 

Repeat the following to reproduce:
1. Run `dsw-tdk put ....`
2. Attempt to render a document

I have attached logs following the commands related to a broken template.

**Docker compose output following `dsw-tdk put ...`**
```
mongo_1       | 2021-02-11T09:46:08.201+0000 I  NETWORK  [listener] connection accepted from 172.23.0.6:46708 #124 (5 connections now open)
dsw_server_1  | [Info ] [I:------------------------------------] [T:f060a545-9e79-4b5b-8cbd-7ade71153d24] [API         ] 201 Created POST /tokens
dsw_server_1  | [Warn ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:------------------------------------] [API         ] 404 Not Found GET /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:23d61224-693e-4082-befe-acd55e3a41c1] [API         ] 201 Created POST /templates
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:df187f65-bde4-4a86-ae22-8e7301efbc04] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:324b6202-ec1a-4036-86ea-34f0ebdb65d5] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:65b10eba-e1c6-443b-adab-f7ccf5830098] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:37cb8e2f-4cb7-4c9d-80fb-1e0f3364ef5a] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:471ef2f1-c891-402f-bb56-1de2c0928dc9] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:ad754e80-5cff-4de3-be7c-b118959a52d2] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:0a3627f7-78f5-4c9e-a041-e5381a9a88ca] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:d90f636b-1499-4878-ba2c-e28a5709bfe9] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:d6cfd258-28e1-47aa-8bc4-69cb81a0b82f] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/assets
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:99aa5f0d-6839-483f-b904-ea42ba0130a7] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:b9b2c20a-0cf6-468b-807c-b7d77bc45350] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:ad940e34-882c-429b-8dde-20fcfd586147] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:db0dc3f0-6dea-4410-aa5d-4e06f2c3fdcc] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:67e05c8c-9386-4077-8926-9b404446a751] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:7140fa48-93a1-43df-be6f-29a71efffe7c] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:47665f33-00f0-47e8-8977-3f4b7435596e] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:83184dc6-c2ca-4020-83fe-aafd03e6b10c] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:c67d031c-467d-4a51-b191-640a8274095b] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:524d4a0f-49aa-418f-b2c8-52bbcac657c6] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:6eeb04e6-e46f-4a79-ad84-0071809e4ee9] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:abdf9b3d-26f9-4e9c-a1c2-a3890a5d8593] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:abb351be-67af-4653-a421-0a41c7fd17f5] [API         ] 201 Created POST /templates/SciLifeLab:SLL-VR-TEMPLATE:1.0.0/files
mongo_1       | 2021-02-11T09:46:09.980+0000 I  NETWORK  [conn124] end connection 172.23.0.6:46708 (4 connections now open)
```

**Docker compose output following attempt to render a document**
```
mongo_1       | 2021-02-11T09:42:44.958+0000 I  NETWORK  [listener] connection accepted from 172.23.0.6:44842 #92 (5 connections now open)
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:2cd1b575-52f5-40b7-a30d-4225c863c834] [Messaging   ] PublishToQueue: document.generation
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:2cd1b575-52f5-40b7-a30d-4225c863c834] [API         ] 201 Created POST /documents
docworker_1   | 2021-02-11 09:42:44,983 | INFO | worker: Received a job
mongo_1       | 2021-02-11T09:42:44.986+0000 I  NETWORK  [conn90] end connection 172.23.0.5:50100 (4 connections now open)
mongo_1       | 2021-02-11T09:42:44.986+0000 I  NETWORK  [conn89] end connection 172.23.0.5:50098 (3 connections now open)
mongo_1       | 2021-02-11T09:42:44.986+0000 I  NETWORK  [listener] connection accepted from 172.23.0.5:50230 #93 (4 connections now open)
docworker_1   | 2021-02-11 09:42:44,988 | INFO | before: Starting call to 'document_worker.worker.Job.connect_mongo', this is the 1st time calling it.
docworker_1   | 2021-02-11 09:42:44,988 | INFO | worker: Connecting to Mongo DB @ mongo:27017/wizard
mongo_1       | 2021-02-11T09:42:44.989+0000 I  NETWORK  [conn93] received client metadata from 172.23.0.5:50230 conn93: { driver: { name: "PyMongo", version: "3.10.1" }, os: { type: "Linux", name: "Linux", architecture: "x86_64", version: "4.19.121-linuxkit" }, platform: "CPython 3.8.7.final.0" }
mongo_1       | 2021-02-11T09:42:44.991+0000 I  NETWORK  [listener] connection accepted from 172.23.0.5:50232 #94 (5 connections now open)
mongo_1       | 2021-02-11T09:42:44.993+0000 I  NETWORK  [conn94] received client metadata from 172.23.0.5:50232 conn94: { driver: { name: "PyMongo", version: "3.10.1" }, os: { type: "Linux", name: "Linux", architecture: "x86_64", version: "4.19.121-linuxkit" }, platform: "CPython 3.8.7.final.0" }
docworker_1   | 2021-02-11 09:42:44,995 | INFO | worker: Getting the document "152f9fc0-48a3-4065-9957-eb5890f1f2fc" details from Mongo DB
docworker_1   | 2021-02-11 09:42:44,996 | INFO | worker: Job "152f9fc0-48a3-4065-9957-eb5890f1f2fc" details received
docworker_1   | 2021-02-11 09:42:44,996 | INFO | worker: Verifying the received job "152f9fc0-48a3-4065-9957-eb5890f1f2fc" details
docworker_1   | 2021-02-11 09:42:44,996 | INFO | worker: Original state of job is QueuedDocumentState
docworker_1   | 2021-02-11 09:42:44,999 | INFO | templates: Loading SciLifeLab:SLL-VR-TEMPLATE:1.0.0
docworker_1   | 2021-02-11 09:42:44,999 | INFO | templates: Verifying SciLifeLab:SLL-VR-TEMPLATE:1.0.0
docworker_1   | 2021-02-11 09:42:44,999 | INFO | templates: Setting up formats for template SciLifeLab:SLL-VR-TEMPLATE:1.0.0
docworker_1   | 2021-02-11 09:42:44,999 | INFO | templates: Storing files of template SciLifeLab:SLL-VR-TEMPLATE:1.0.0 locally
docworker_1   | 2021-02-11 09:42:45,002 | INFO | templates: Storing assets of template SciLifeLab:SLL-VR-TEMPLATE:1.0.0 locally
docworker_1   | 2021-02-11 09:42:45,002 | INFO | formats: Setting up format "PDF Document" (45469a71-272c-4e3d-833e-cf6492cbcf17)
docworker_1   | 2021-02-11 09:42:45,012 | INFO | worker: Building document by rendering template with context
docworker_1   | 2021-02-11 09:42:45,012 | ERROR | worker: (152f9fc0-48a3-4065-9957-eb5890f1f2fc) Failed to build final document: [TemplateNotFound] parts/macros.html.j2
docworker_1   | 2021-02-11 09:42:45,014 | INFO | worker: (152f9fc0-48a3-4065-9957-eb5890f1f2fc) Set state to ErrorDocumentState
docworker_1   | 2021-02-11 09:42:45,014 | INFO | worker: Job processing finished (ACK sent)
```

**Docker compose output following attempt to delete template**
```
dsw_server_1  | [Info ] [I:3ef96067-8e7a-479e-ae18-1be47f0e2a5d] [T:d4ae7951-60a5-48ca-8c6f-3b8fe5bb4a28] [API         ] 204 No Content DELETE /templates/?organizationId=SciLifeLab&templateId=SLL-VR-TEMPLATE

```

**Comment by MarekSuchanek** [comment]
OK, I was finally able to reproduce it. It is caused by the asynchronous upload... as it call several "edit" on one entity, it may cause some overwrites (the more files you have the more probability that it will happen).

It will be fixed in 2.13.0 (JIRA task: [DSW-1043](https://ds-wizard.atlassian.net/browse/DSW-1043)).

Thanks for reporting 👍

**Comment by MarekSuchanek** [comment]
`v2.13.0` released, please check and re-open if necessary
